### PR TITLE
LINK-1989 | ICS support for recurring events

### PR DESCRIPTION
**Pull request by harriris-vincit**
### Description
Adds support for creating calendar events for recurring event's sub-events and appending those to the ICS calendar.
### Closes
[LINK-1989](https://helsinkisolutionoffice.atlassian.net/browse/LINK-1989)

[LINK-1989]: https://helsinkisolutionoffice.atlassian.net/browse/LINK-1989?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr938.api.dev.hel.ninja** :rocket::rocket::rocket:

File: registrations/utils.py [hunk 1]
```diff
@@ -108,37 +108,50 @@ def has_allowed_substitute_user_email_domain(email_address):
...
+    local_tz = pytz.timezone(settings.TIME_ZONE)
 
-        if (start_time := event.start_time) and (name := event.name):
-            calendar_event.add("dtstart", start_time)
-            calendar_event.add("summary", name)
+    with translation.override(language):
+        if event.is_recurring_super_event:
+            for sub_event in event.sub_events.filter(deleted=False):
+                calendar_event = _create_calendar_event_from_event(sub_event, local_tz)
+                cal.add_component(calendar_event)
```
Comment (jorilindell) [inline registrations/utils.py:151]: Should we still generate ICS file if there is at least one valid calender_event?

Admin user can e.g. postpone one of the sub events and in that case the start_time of that particular sub event is set to null. If the read the code correctly the ics file generation will fail in that case.
 
Screenshot 2024-06-14 at 11 22 41

Comment (jorilindell) [reply registrations/utils.py:151]: How about cancelled events (event_status = "EventCancelled")? Do we want to include those events in the calendar file as well or should they be filtered out.

Comment (nicobav) [reply registrations/utils.py:151]: Nice catch 🎣 
Indeed raising a question should the whole series be PUBLISHED or do we just need to care about the published ones. 🤔

Comment (harriris-vincit) [reply registrations/utils.py:151]: Nice catch indeed. PO opinion needed here probably so I'll ask aroun in Slack.

Comment (harriris-vincit) [reply registrations/utils.py]: Changed this to only create the ICS file for sub-events if the sub-events are published, scheduled and not soft-deleted.

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr938.api.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr938.api.dev.hel.ninja** :rocket::rocket::rocket:

File: registrations/models.py [hunk 2]
```diff
@@ -660,6 +660,18 @@ def send_event_cancellation_notifications(self, is_sub_event_cancellation=False)
                 is_sub_event_cancellation=is_sub_event_cancellation,
             )
 
+    def get_calendar_events(self):
+        if self.event.is_recurring_super_event:
+            events = self.event.sub_events.filter(
+                deleted=False,
+                event_status=Event.Status.SCHEDULED,
```
Comment (jorilindell) [inline registrations/models.py:667]: I think that both Event.Status.SCHEDULED and Event.Status.RESCHEDULED should be included.

Comment (harriris-vincit) [reply registrations/models.py]: Added `RESCHEDULED` status as well.

**Review by jorilindell — APPROVED** [review]
Looks very good!

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr938.api.dev.hel.ninja** :rocket::rocket::rocket:

**Pull request merged**